063-topic-replication-factor-change.md

```diff
@@ -0,0 +1,240 @@
+    "intraBrokerDataToMoveMB": 0,
+    "monitoredPartitionsPercentage": 100,
+    "provisionRecommendation": "",
+    "excludedBrokersForReplicaMove": [],
+    "excludedBrokersForLeadership": [],
+    "provisionStatus": "RIGHT_SIZED",
+    "onDemandBalancednessScoreBefore": 100,
+    "recentWindows": 5,
+    "dataToMoveMB": 0,
+    "excludedTopics": [],
+    "numLeaderMovements": 0
+  }
+  ...
+```
+
+The `goals` parameter is the comma separated list of goals used to generate the configuration proposal.
+This is configurable for all endpoints in `Kafka.spec.cruiseControl.config` with `default.goals` property.
+
+The `replication_throttle` parameter is an upper bound on the bandwidth used to move replicas (bytes per second).
+This is configurable for all endpoints in `Kafka.spec.cruiseControl.config` with `default.replication.throttle` property.
+
+The `skip_rack_awareness_check` boolean parameter is undocumented and disabled by default.
+In case of a single-rack deployment, rack goals are not configured and the Topic Operator will automatically enable `skip_rack_awareness_check`.
+
+#### User tasks endpoint
+
+The `user_tasks` endpoint is used to periodically check the tasks state by `User-Task-ID`, which is returned in the `topic_configuration` response.
+
+```sh
+$ curl -s "localhost:9090/kafkacruisecontrol/user_tasks?user_task_ids=5344ca89-351f-4565-8d0f-9aade00e053d&json=true" | jq
+{
+  "userTasks": [
+    {
+      "Status": "Completed",
+      "ClientIdentity": "127.0.0.1",
+      "RequestURL": "POST /kafkacruisecontrol/topic_configuration?topic=my-topic&replication_factor=2&skip_rack_awareness_check=true&dryrun=false&json=true",
+      "UserTaskId": "5344ca89-351f-4565-8d0f-9aade00e053d",
+      "StartMs": "1701961135978"
+    }
+  ],
+  "version": 1
+}
+```
+
+### Topic Operator
+
+At runtime, the Topic Operator watches and periodically reconciles the configuration of all managed and unpaused KafkaTopic resources.
+Among the other configurations, the reconciliation logic also detects `.spec.replicas` changes by comparing with the actual topic replication factor in Kafka.
+From the point of view of the Topic Operator, a topic replication factor change can be in the following states:
+
+- **Pending**: replicas is updated, but the replication factor change request is not sent to or accepted by Cruise Control.
+- **Ongoing**: replicas is updated and Cruise Control has accepted the request (task queued), but the execution is not completed (possible task states: Active and InProgress).
+- **Ready**: replicas is updated and Cruise Control task execution is completed (possible task states: Completed).
+
+When Cruise Control integration is enabled and a replicas change is detected by the BatchingTopicController, the event is notified to the ReplicasChangeManager.
+The ReplicasChangeManager maintains a cache of pending and ongoing replicas changes, manages the communication with Cruise Control's REST API and updates the KafkaTopic status.
```

Comment:
You keep ignoring my comments and questions on this one. I'm missing an explanation why this is needed - the `ReplicasChangeManager` and its cache ... or how will it work. My assumption is that the easiest way to handle this is in the regular reconciliation:
1. You check the status if there is already a CC task (and either handle it if is it complete and remove it from status or just check it and keep going if it is still queued or in progress)
2. If there is no task in the status, you check if the replicas in Kafka and in the KafkaTopic resource are the same. And if not, you _try to create the task in CC_.

What do you need the `ReplicasChangeManager` and its cache for and how will it work? Maybe there is some reason, but it seems you are not really explaining it anywhere and just closing my comments.

Reply:
### Isolate complexity

It is always the regular reconciliation that drives the ReplicasChangeManager, by providing replicas change events. The Cruise Control REST API integration is quite complex with different HTTP requests and polling. Using a dedicated object seems to be a natural choice to isolate that complexity from the rest of the reconciliation logic. This is somewhat similar to the choice of providing a Reconciler object where a single component reconciliation logic is hosted.

### Decouple asynchronous reconciliation

A replicas change reconciliation is very different from the reconciliation of other topic configurations, as it is executed asynchronously. After a replicas change event is notified, it takes some time for the request to land in Cruise Control's queue, and some more time to be actually executed. Most reconciliations will find that the replicas change is still pending or ongoing, due to the limited Cruise Control capacity (only one task can be executed at a time). This would generate a lot of unnecessary HTTP requests, that would slowdown the reconciliation logic for no good. 

### Avoid duplicates and enable batch mode

The ReplicasChangeManager's cache is needed to avoid generating duplicated requests that would overload Cruise Control and fill the task queue with unnecessary tasks, with the effect of delaying valid replicas change requests. It is also used to group multiple replicas changes together and map them to a single Cruise Control request/task (batch mode). This is useful in case if bulk replicas change (i.e. gitops update).

Let me know if it is clear. 
Will add these points to the proposal.

Reply:
Thanks for providing the reasoning. While I might not agree with it in all cases, I think it is definitely useful to have it in the proposal as an explanation. That said, this does not explain anything about how will the `ReplicasChangeManager` work. All I have are some guesses based on some parts of it you mention such as the cache etc. So it would be good to cover that as well.

----

> Isolate complexity

I do not think nobody prevents you from having classes to abstract the CC API Cruise Control complexity. But I do not think you provide any reasons why to have a _separate operator_ for it. The polling can be easily done from the regular reconciliations. Or maybe you badly explain how will the `ReplicasChangeManager` work.

The Reconciler classes in the Cluster Operator - if that is what you were referring to - are stateless. They do not keep any state (cache) and are called only from resource reconciliation. I think that is a valid pattern but your reasoning does not seem to suggest this is the same as what you are planing.

> Decouple asynchronous reconciliation

I do not understand the promised _efficiency. The reconciliation of a KafkaTopic CR happens normally every 2? minutes (Or does TO still use 90 seconds?) and on changes. That does not seem to overload any CC APIs. How often will the `ReplicasChangeManager` call it? Or how would it reduce the API calls?

> Avoid duplicates and enable batch mode

Duplicates can be avoided when driving this through the custom resources as well => at the end this is the backup for operator restarts. You either have the task ID you got stored in it or not. Batching in my mind does not justify the complexity. Do you really expect bulk changes in the replication factor to be a common thing? What would be the use-cases? _Latency_ of how quickly the changes are applied also does not seem to be a major issue here.

Reply:
Let me answer some questions, then I'll try to add more details to the proposal.

> The reconciliation of a KafkaTopic CR happens normally every 2? minutes (Or does TO still use 90 seconds?) and on changes. That does not seem to overload any CC APIs. 

The backlog of changes may be big not only after a bulk update, but also when CC is running a long rebalance task. In this case, all replicas changes have to wait and accumulate in the TO. Assuming no batching, this would generate 2 HTTP requests for each replicas change on every periodic reconciliation.

> How often will the ReplicasChangeManager call it? Or how would it reduce the API calls?

The ReplicasChangeManager calls the configuration endpoint just one time, with all unknown replicas changes. If there a no new replicas changes, no request is sent. Then, it polls the tasks endpoint every 10 seconds passing all ongoing changes (one request). In case of any error, it will wait for 60 seconds and then retry, indefinitely.

> Duplicates can be avoided when driving this through the custom resources as well => at the end this is the backup for operator restarts.

That's true. I'll try to see if can filter out pending changes using the ReplicasChanged condition and get rid of the pendingChanges cache, and only keep ongoing as an optimization to avoid calling Kube API every 10 seconds when polling for status.

Reply:
It is obviously subjective, but I think you overestimate how much will this be used. The replication factor changes are super painful because of how complicated the UX of the Kafka tool is (I guess the fact nobody fixed the UX also shows how often it is used). But are IMHO not done very often. If you believe the RF to be changed all the time up and down, I also wonder how will you batch the requests (will there be a time window? Will there be a size limit?) and deal with the memory consumption of your cache.

I personally would go for simplicity, but it is your proposal. So if you wanna stick with it, just make sure it is properly described. 

> That's true. I'll try to see if can filter out pending changes using the ReplicasChanged condition, getting rid of the pendingChanges cache, and only keep the ongoing cache as an optimization to avoid calling the Kube API every 10 seconds when polling for status.

I do not understand why would you call Kubernetes API every 10 seconds. But you should definitely not do that. As I said before, you should add an explanation how will this work.

Reply:
> I do not understand why would you call Kubernetes API every 10 seconds.

In fact, I don't do it.

Reply:
I'm making few changes to make the ReplicasChangeManager completely driven by the regular reconciliation. Then, I'll update the proposal.